promql/engine.go

```diff
@@ -1325,16 +1421,111 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.interval = ev.noStepSubqueryIntervalFn(rangeMillis)
 		}
 
-		// Start with the first timestamp after (ev.startTimestamp - offset - range)
+		subqueryStartTime := ev.startTimestamp - offsetMillis - rangeMillis
+		if e.Timestamp != 0 {
+			newEv.endTimestamp = e.Timestamp - offsetMillis
+			subqueryStartTime = newEv.endTimestamp - rangeMillis
+		}
+
+		// Start with the first timestamp after subqueryStartTime
 		// that is aligned with the step (multiple of 'newEv.interval').
-		newEv.startTimestamp = newEv.interval * ((ev.startTimestamp - offsetMillis - rangeMillis) / newEv.interval)
-		if newEv.startTimestamp < (ev.startTimestamp - offsetMillis - rangeMillis) {
+		newEv.startTimestamp = newEv.interval * (subqueryStartTime / newEv.interval)
+		if newEv.startTimestamp < subqueryStartTime {
 			newEv.startTimestamp += newEv.interval
 		}
 
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
 		return res, ws
+	case *parser.StepInvariantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
+		}
+		newEv := &evaluator{
+			startTimestamp:           ev.startTimestamp,
+			endTimestamp:             ev.startTimestamp,
+			interval:                 ev.interval,
+			ctx:                      ev.ctx,
+			currentSamples:           ev.currentSamples,
+			maxSamples:               ev.maxSamples,
+			logger:                   ev.logger,
+			lookbackDelta:            ev.lookbackDelta,
+			noStepSubqueryIntervalFn: ev.noStepSubqueryIntervalFn,
+		}
+
+		isSelector := false
+		switch e.Expr.(type) {
+		case *parser.VectorSelector, *parser.MatrixSelector, *parser.SubqueryExpr:
+			isSelector = true
+		}
+		if isSelector {
```

Comment:
Can't this produce multiple steps worth of results then? Doing one step and then duplicating would be cheaper.

You also still look to be creating an evaluator per node.

Reply:
> Can't this produce multiple steps worth of results then?

I will check about that, since vector selector and matrix selector select at some constant time, I dont think it will evaluate for multiple steps. And I don't see a requirement or possibility of duplication for matrix selector and subqueries.

> You also still look to be creating an evaluator per node.

Only the top most node upto which it is step invariant

Reply:
> I will check about that, since vector selector and matrix selector select at some constant time, I dont think it will evaluate for multiple steps.

That doesn't sound right, given you're specifying a non-zero range for the query here.

> Only the top most node upto which it is step invariant

That's still not what I'm seeing in the code.

Reply:
> That doesn't sound right, given you're specifying a non-zero range for the query here.

Updated now, does not require a non 0 time range now

> That's still not what I'm seeing in the code.

I have added `TestWrapWithStepInvariantExpr` in engine_test to show the wrapping. You can see that it is wrapped for the top most level possible in the AST.

Reply:
The tests do not demonstrate that, try a nested function or more than one operator to see why. E.g. abs(abs(a@1234))

Reply:
Adding more test cases for the wrapping

Reply:
Did you forget to push?

Reply:
I will be pushing soon, trying to combine multiple reviews in a single commit.